src/Server/TCPHandler.cpp

```diff
@@ -1874,7 +1875,13 @@ void TCPHandler::sendHello()
 
     if (client_tcp_protocol_version >= DBMS_MIN_REVISION_WITH_SERVER_SETTINGS)
     {
-        if (is_interserver_mode || !Context::getGlobalContextInstance()->getServerSettings()[ServerSetting::send_settings_to_client])
+        if (is_interserver_mode ||
+            !Context::getGlobalContextInstance()->getServerSettings()[ServerSetting::send_settings_to_client] ||
+            /// If the client is new enough to receive settings but not new enough to know about
+            /// apply_settings_from_server setting, we effectively apply this setting on the server
+            /// side, right here.
+            (client_tcp_protocol_version < DBMS_MIN_REVISION_WITH_APPLY_SETTINGS_FROM_SERVER_SETTING
+                && !session->sessionContext()->getSettingsRef()[Setting::apply_settings_from_server]))
```

Comment:
Why this extra check for `apply_settings_from_server` is required?

Reply:
To support clients built after https://github.com/ClickHouse/ClickHouse/pull/71178 but before this PR. If we set `compatibility` setting to < 24.12 in server's users.xml, and such client connects to the server, it seems better to not send the settings. This way the behavior would be the same across all client versions (old, intermediate, new) (as long as `apply_settings_from_server` is not overridden by the query or similar) - they all won't apply server settings.